Add EntityHarvestBlockEvent

Doc94 opened the pull request:
This starts in [discord](https://canary.discord.com/channels/289587909051416579/555462289851940864/1327837651876970569) where the Composter when drop a bone meal not has any event... the first thing was use the Drop or Dispenser event but that not handle correctly in this behaviour... then Machine Maker suggest a new event.

Then this PR add a new event for the case when a Entity harvest a block currently for Vines and Composter. this new event cannot be a base of the player side by the original design... then add in the docs the mention for this cases...

Doc94 committed:
Add EntityHarvestBlockEvent

Warriorrrr commented on paper-api/src/main/java/org/bukkit/event/player/PlayerHarvestBlockEvent.java:
```diff
@@ -5,13 +5,16 @@
 /**
  * This event is called whenever a player harvests a block.
  * <br>
+ * For the entity case please use {@link org.bukkit.event.entity.EntityHarvestBlockEvent}
```
I think this line here and in the new event's javadoc could be a bit better, since there's an import it's also not necessary to use the fully qualified name.
```suggestion
 * For cases involving entities, please use {@link EntityHarvestBlockEvent}.
```

Warriorrrr commented on paper-api/src/main/java/org/bukkit/event/entity/EntityHarvestBlockEvent.java:
```diff
@@ -0,0 +1,76 @@
+package org.bukkit.event.entity;
```
New events should go into the paper package

Warriorrrr commented on paper-api/src/main/java/org/bukkit/event/entity/EntityHarvestBlockEvent.java:
```diff
@@ -0,0 +1,76 @@
+ */
+public class EntityHarvestBlockEvent extends EntityEvent implements Cancellable {
+
+    private static final HandlerList handlers = new HandlerList();
```
Should be HANDLER_LIST

Warriorrrr commented on paper-server/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java:
```diff
@@ -344,11 +345,22 @@ public static EntityEnterLoveModeEvent callEntityEnterLoveModeEvent(net.minecraf
      */
     public static PlayerHarvestBlockEvent callPlayerHarvestBlockEvent(Level world, BlockPos blockposition, net.minecraft.world.entity.player.Player who, InteractionHand enumhand, List<ItemStack> itemsToHarvest) {
-        List<org.bukkit.inventory.ItemStack> bukkitItemsToHarvest = new ArrayList<>(itemsToHarvest.stream().map(CraftItemStack::asBukkitCopy).collect(Collectors.toList()));
+        List<org.bukkit.inventory.ItemStack> bukkitItemsToHarvest = itemsToHarvest.stream().map(CraftItemStack::asBukkitCopy).collect(Collectors.toList());
```
While this does work I think it's important to note that Collectors#toList doesn't guarantee that it's mutable, but since that's unlikely to change it's probably fine.

Doc94 committed:
Docs updates for PlayerHarvestBlockEvent

Co-authored-by: Warrior <50800980+Warriorrrr@users.noreply.github.com>

Doc94 committed:
Move EntityHarvestBlockEvent to paper package and HandlerList change

# Conflicts:
#	paper-server/patches/sources/net/minecraft/world/level/block/CaveVines.java.patch
#	paper-server/patches/sources/net/minecraft/world/level/block/ComposterBlock.java.patch
#	paper-server/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java

Doc94 committed:
[ci skip] Make EntityHarvestBlockEvent follow the annotation standard

Warriorrrr commented on paper-api/src/main/java/io/papermc/paper/event/entity/EntityHarvestBlockEvent.java:
```diff
@@ -0,0 +1,77 @@
+     *
+     * @return A list of items that are being harvested from this block
+     */
+    public List<ItemStack> getItemsHarvested() {
```
It could be noted here and in the player event that this list is mutable.

Warriorrrr reviewed (changes requested):
The case of foxes harvesting a sweet berry bush is currently missing, can be added in the FoxEatBerriesGoal#pickSweetBerries method

Doc94 committed:
[ci skip] improvement in docs for mutable list

notTamion commented on paper-api/src/main/java/io/papermc/paper/event/entity/EntityHarvestBlockEvent.java:
```diff
@@ -0,0 +1,77 @@
+     *
+     * @return The block that is being harvested
+     */
+    public Block getHarvestedBlock() {
+        return harvestedBlock;
+    }
+
+    /**
+     * Gets a list of items that are being harvested from this block.
+     *
+     * @return A mutable list of items that are being harvested from this block
+     */
+    public List<ItemStack> getItemsHarvested() {
```
i would maybe add that for foxes the first item will be used for its mouth

Doc94 replied:
Like apiNote?

notTamion replied:
yeah or just normally in the comment. both is fine with me

notTamion commented on paper-server/patches/sources/net/minecraft/world/level/block/ComposterBlock.java.patch:
```diff
@@ -48,6 +48,33 @@
          if (!level.isClientSide) {
              Vec3 vec3 = Vec3.atLowerCornerWithOffset(pos, 0.5, 1.01, 0.5).offsetRandom(level.random, 0.7F);
              ItemEntity itemEntity = new ItemEntity(level, vec3.x(), vec3.y(), vec3.z(), new ItemStack(Items.BONE_MEAL));
+             itemEntity.setDefaultPickUpDelay();
```
can also remove this call as we call it later anyway

notTamion commented on paper-server/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java:
```diff
@@ -344,11 +345,22 @@ public static EntityEnterLoveModeEvent callEntityEnterLoveModeEvent(net.minecraf
+        List<org.bukkit.inventory.ItemStack> bukkitItemsToHarvest = itemsToHarvest.stream().map(CraftItemStack::asBukkitCopy).collect(Collectors.toList());
+        org.bukkit.entity.Entity entity = who.getBukkitEntity();
+        EntityHarvestBlockEvent entityHarvestBlockEvent = new EntityHarvestBlockEvent(entity, CraftBlock.at(world, blockposition), bukkitItemsToHarvest);
+        Bukkit.getPluginManager().callEvent(entityHarvestBlockEvent);
```
```suggestion
        entityHarvestBlockEvent.callEvent();
```

notTamion commented on paper-server/patches/sources/net/minecraft/world/entity/animal/Fox.java.patch:
```diff
@@ -130,11 +130,41 @@
              int i = 1 + Fox.this.level().random.nextInt(2) + (ageValue == 3 ? 1 : 0);
              ItemStack itemBySlot = Fox.this.getItemBySlot(EquipmentSlot.MAINHAND);
++            // Paper start - call HarvestBlockEvent for the item dropped
++            ItemStack itemHarvest = new ItemStack(Items.SWEET_BERRIES, i);
```
Haven't tested but just a thought to make sure. considering in the vanilla logic we would popresource i times that would produce multiple items on the ground that are not immediately bundled together right? though since we just creating one itemstack with i amount here those would all immediately be bundled together when they get spawned. not sure but just going to leave it here so we don't overlook it

Doc94 replied:
maybe i bad read this but the logic is the same than vanilla, the event is called with only one item and is used for pop, if you add more items then pop for every element.. the only thing is if you make a thing like "add potato", "add potato" and that are different items added in the list... but still respect the original thing.

notTamion replied:
oh yeah nvm i read the original if statement as a while statement

notTamion commented on paper-server/patches/sources/net/minecraft/world/level/block/ComposterBlock.java.patch:
```diff
@@ -48,6 +48,33 @@
++            java.util.ArrayList<org.bukkit.inventory.ItemStack> itemsHarvested = new java.util.ArrayList<>();
++            if (entity instanceof Player player) {
++                org.bukkit.event.player.PlayerHarvestBlockEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerHarvestBlockEvent(level, pos, player, InteractionHand.MAIN_HAND, java.util.Collections.singletonList(itemEntity.getItem()));
++                itemsHarvested.addAll(event.getItemsHarvested());
```
no need to clone the list here since we aren't modifying it later.

Doc94 committed:
[ci skip] mention Fox behaviour for EntityHarvestBlockEvent#getItemsHarvested()

notTamion committed:
clarify Fox behaivour

notTamion reviewed (approved):
LGTM

Warriorrrr reviewed (changes requested):
Looking back at this now it would indeed seem better to make the new event more proper with defensive copying for the list and items and adding setters.

Doc94 committed:
Fix Conflicts mess git

Doc94 committed:
Clone Items and make immutable
